src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaUserDetailsContextMapper.java

```diff
@@ -0,0 +1,325 @@
+     * @param username User name to assign to local User, used for signing in
+     * @return User
+     */
+    private User createNewUserFromLdap(DirContextOperations dirContextOperations, String username) {
+        logger.info("Creating new IRIDA user for found LDAP user: " + username);
+        Locale locale = LocaleContextHolder.getLocale();
+
+        User u = ldapCreateUser(dirContextOperations, username);
+
+        try {
+            // Commit user to the database
+            commitUser(u);
+        } catch (EntityExistsException e) {
+            String ldap_error1 = messageSource.getMessage(
+                    "LoginPage.ldap_error.description_1", null, locale);
+            logger.error(ldap_error1 + e);
+            throw new IridaLdapAuthenticationException(ldap_error1, e, 1);
+        } catch (ConstraintViolationException e) {
+            String ldap_error2 = messageSource.getMessage(
+                    "LoginPage.ldap_error.description_2", null, locale);
+            logger.error(ldap_error2 + e);
+            throw new IridaLdapAuthenticationException(ldap_error2, e, 2);
+        }
+
+        try {
+            // return the newly created user
+            return userRepository.loadUserByUsername(username);
+        }
+        catch(UsernameNotFoundException e) {
+            String ldap_error3 = messageSource.getMessage(
+                    "LoginPage.ldap_error.description_3", null, locale);
+            logger.error(ldap_error3 + e);
+            throw new IridaLdapAuthenticationException(ldap_error3, e, 3);
+        }
+    }
+
+    /**
+     *
+     * @param dirContextOperations Ldap/adLdap service context
+     * @param username username used for sign-in
+     * @return new {@link User} object
+     */
+    private User ldapCreateUser(DirContextOperations dirContextOperations, String username) {
+        // This works for both ldap and adLdap
+        String randomPassword = generateCommonLangPassword();
+        Map<String, String> map = getLdapFields(dirContextOperations);
+        return new User(
+                username,
+                map.get(iridaUserFieldEmail),
+                randomPassword,
+                map.get(iridaUserFieldFirstName),
+                map.get(iridaUserFieldLastName),
+                map.get(iridaUserFieldPhoneNumber)
+        );
+    }
+
+    /**
+     * Fetch data from LDAP/ADLDAP and create a map that aligns with {@link User}
+     * @param dirContextOperations LDAP/ADLDAP context
+     * @return Map<String {@link User} property, String LDAP/ADLDAP field value>
+     */
+    private Map<String, String> getLdapFields(DirContextOperations dirContextOperations) {
+        Map<String, String> map = new HashMap<>();
+        map.put(iridaUserFieldEmail, getAttribute(dirContextOperations, userInfoEmail, true, ""));
+        map.put(iridaUserFieldFirstName, getAttribute(dirContextOperations, userInfoFirstName, false, "FirstName"));
```

Comment:
Should we really have a fallback for `firstName` and `lastName`, I lean towards no since this would be very confusing to see the list of users in IRIDA since these won't be modifiable.

Reply:
Good catch. I have changed the name fields fields to be required now.